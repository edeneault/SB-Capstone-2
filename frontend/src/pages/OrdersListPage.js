import React, { useEffect } from "react";
import { LinkContainer } from "react-router-bootstrap";
import { Container, Table, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import Loader from "../components/Loader";
import { listOrders } from "../actions/orderActions";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const UserListScreen = ({ history }) => {
  const dispatch = useDispatch();

  const ordersList = useSelector((state) => state.ordersList);
  const { loading, error, orders } = ordersList;

  const userLogin = useSelector((state) => state.userLogin);
  const { userInfo } = userLogin;

  useEffect(() => {
    if (userInfo && userInfo.isAdmin) {
      dispatch(listOrders());
    } else {
      history.push("/login");
    }
  }, [dispatch, history, userInfo]);

  return (
    <>
      <Container fluid className='px-5'>
        <h1 className='page-title slide-in'>My Orders</h1>
        {loading ? (
          <Loader />
        ) : error ? (
          toast("Something went wrong", { type: "error" })
        ) : (
          <Table bordered hover responsive className='table-sm'>
            <thead>
              <tr className='bg-black text-light'>
                <th>ORDER ID</th>
                <th>USER</th>
                <th>DATE</th>
                <th>TOTAL</th>
                <th>PAID</th>
                <th>DELIVERED</th>
                <th>ACTION</th>
              </tr>
            </thead>
            <tbody>
              {orders.map((order) => (
                <tr key={order._id} className='align-middle'>
                  <td>{order._id}</td>
                  <td>{order.user && order.user.name}</td>
                  <td>{order.createdAt}</td>
                  <td>{order.totalPrice}</td>

                  <td className='text-center'>
                    {order.isPaid ? (
                      order.paidAt.substring(0, 10)
                    ) : (
                      <i className='fas fa-times' style={{ color: "red" }}></i>
                    )}
                  </td>

                  <td className='text-center'>
                    {order.isDelivered ? (
                      order.deliveredAt.substring(0, 10)
                    ) : (
                      <i className='fas fa-times' style={{ color: "red" }}></i>
                    )}
                  </td>
                  <td>
                    <LinkContainer to={`/orders/${order._id}`}>
                      <Button variant='light' className='btn-sm'>
                        DETAILS
                      </Button>
                    </LinkContainer>
                  </td>
                </tr>
              ))}
            </tbody>
          </Table>
        )}
      </Container>
    </>
  );
};

export default UserListScreen;
