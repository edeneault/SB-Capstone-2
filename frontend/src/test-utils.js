import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
