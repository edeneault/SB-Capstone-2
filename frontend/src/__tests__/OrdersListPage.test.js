import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { initialState, render } from "../test-utils";
import OrdersListPage from "../pages/OrdersListPage.js";
import { createMemoryHistory } from "history";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  const history = createMemoryHistory();

  test("renders OrdersListPage without crashing", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <OrdersListPage history={history} />
      </Provider>,
      { initialRoutes: [`/admin/orderslist`] },
    );
    // screen.debug();
    expect(getByText(/My Orders/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <OrdersListPage history={history} loading={false} />
      </Provider>,
      { initialRoutes: [`/admin/orderslist`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
