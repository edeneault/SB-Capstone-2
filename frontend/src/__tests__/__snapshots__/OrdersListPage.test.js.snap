// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="fade-in px-5 pt-3 container-fluid"
  >
    <h1
      class="page-title slide-in"
    >
      My Orders
    </h1>
    <div
      class="table-responsive"
    >
      <table
        class="table-sm table table-bordered table-hover"
      >
        <thead>
          <tr
            class="bg-black text-light"
          >
            <th>
              ORDER ID
            </th>
            <th>
              USER
            </th>
            <th>
              DATE
            </th>
            <th>
              TOTAL
            </th>
            <th>
              PAID
            </th>
            <th>
              DELIVERED
            </th>
            <th>
              ACTION
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="align-middle"
          >
            <td>
              6125d0c2d3ffd29bedee24af
            </td>
            <td>
              AdminName
            </td>
            <td>
              2021-08-25T05:10:26.957Z
            </td>
            <td>
              459.99
            </td>
            <td
              class="text-center"
            >
              2021-08-26
            </td>
            <td
              class="text-center"
            >
              2021-08-26
            </td>
            <td>
              <a
                class="btn-sm btn btn-light"
                href="/orders/6125d0c2d3ffd29bedee24af"
              >
                DETAILS
              </a>
            </td>
          </tr>
          <tr
            class="align-middle"
          >
            <td>
              6126a23b644a96d02caece4c
            </td>
            <td>
              AdminName
            </td>
            <td>
              2021-08-25T20:04:11.815Z
            </td>
            <td>
              459.99
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-times"
                style="color: red;"
              />
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-times"
                style="color: red;"
              />
            </td>
            <td>
              <a
                class="btn-sm btn btn-light"
                href="/orders/6126a23b644a96d02caece4c"
              >
                DETAILS
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
